xilem/examples/stopwatch.rs

```diff
@@ -0,0 +1,261 @@
+use xilem::{WidgetView, Xilem};
+use xilem_core::fork;
+use xilem_core::one_of::Either;
+
+/// The state of the entire application.
+///
+/// This is owned by Xilem, used to construct the view tree, and updated by event handlers.
+struct Stopwatch {
+    active: bool,
+    /// The duration to add to the duration since the last instant.
+    /// This is needed since you can pause a timer, and we need to account for all
+    /// time the timer was active before the last start.
+    added_duration: Duration,
+    /// The absolute time of the last start for calculating elapsed time.
+    last_start_time: Option<SystemTime>,
+    /// The duration displayed; updated by by `update_display()`
+    displayed_duration: Duration,
+    /// An error string to display if there is an error.
+    displayed_error: String,
+    /// A list of the length of all completed splits. Does not include the current split.
+    completed_lap_splits: Vec<Duration>,
+    /// The duration of the main timer when the split was started.
+    split_start_time: Duration,
+}
+
+impl Stopwatch {
+    fn start(&mut self) {
+        self.last_start_time = Some(SystemTime::now());
+        self.active = true;
+        self.update_display();
+    }
+
+    fn stop(&mut self) {
+        let dur_since_last_instant = self
+            .last_start_time
+            .expect("stop should only be called when the start time is set")
+            .elapsed();
+        match dur_since_last_instant {
+            Ok(dur) => {
+                self.added_duration = self.added_duration.add(dur);
+                self.displayed_error = "".into();
+            }
+            Err(err) => {
+                self.displayed_error = format!("failed to calculate elapsed time: {}", err);
+            }
+        }
+        self.last_start_time = None;
+        self.active = false;
+        self.update_display();
+    }
+
+    fn reset(&mut self) {
+        self.active = false;
+        self.last_start_time = None;
+        self.added_duration = Duration::ZERO;
+        self.completed_lap_splits.clear();
+        self.split_start_time = Duration::ZERO;
+        self.update_display();
+    }
+
+    fn lap(&mut self) {
+        let split_end = self.get_current_duration();
+        self.completed_lap_splits
+            .push(split_end.sub(self.split_start_time));
+        self.split_start_time = split_end;
+    }
+
+    fn get_current_duration(&self) -> Duration {
+        match self.last_start_time {
+            Some(last_start_time) => match last_start_time.elapsed() {
+                Ok(elapsed) => self.added_duration + elapsed,
+                Err(err) => {
+                    warn!("error calculating elapsed time: {}", err.to_string());
+                    self.added_duration
+                }
+            },
+            _ => self.added_duration,
+        }
+    }
+
+    fn update_display(&mut self) {
+        self.displayed_duration = self.get_current_duration();
```

Comment:
Can we just call `self.get_current_duration()` whenever `self.displayed_duration` would be used?

Reply:
I don't think that would work right. The timestamp doesn't change because it's an absolute time. The async repeat call needs to trigger a change for the label to detect.